### PR TITLE
Optimized Traversable<>.unzip(Function, Function)

### PR DESCRIPTION
**Pull request by sleepytomcat**
New unzip() method accepts two functions to avoid extensive Tuple creation under the hood.

### CODE REVIEW

**Comment by sleepytomcat** [comment]
This is a PR to address https://github.com/vavr-io/vavr/issues/2503

File: src/main/java/io/vavr/collection/Iterator.java
Comment (sleepytomcat) [inline src/main/java/io/vavr/collection/Iterator.java:1222]: This one looks particularly questionable, as we do create a Stream of Tuples here, and then split them into Tuple of Iterators. However I could not find a better solution yet, as Iterator has mutable state, and generally traversable only once.

Comment (sleepytomcat) [reply src/main/java/io/vavr/collection/Iterator.java:1222]: Are there better ways to do this?

Comment (danieldietrich) [reply src/main/java/io/vavr/collection/Iterator.java:1222]: Right, I would do it as follows:

```java
// ...
final Iterator<T1> iter1 = map(unzipper1);
final Iterator<T2> iter2 = map(unzipper2);          
return Tuple.of(iter1, iter2);
```

**Review by danieldietrich — CHANGES_REQUESTED** [review]
Hi Sergei,

many thanks for your PR!
I've thought a bit about the unzip method. I think we should provide only one implementation in `Traversable`:

```java
default <T1, T2> Tuple2<Iterator<T1>, Iterator<T2>> unzip(...) {
    Objects.requireNonNull(unzipper1, "unzipper1 is null");
    Objects.requireNonNull(unzipper2, "unzipper2 is null");
    final Iterator<T1> iter1 = iterator().map(unzipper1);
    final Iterator<T2> iter2 = iterator().map(unzipper2);
    return Tuple.of(iter1, iter2);
}
```

And we will have the additional implementation in `Iterator`, like described below.

The reasons for the central implementation are:

1. There emerge problems in sorted collections because we do not know, if the target types T1, T2 are comparable. That might lead to ClassCastExceptions at runtime.
2. I do not want to be too restrictive. The users should decide on their own which collection they target. The Tuple2 of Iterators allows them to choose their collection later. They even could proceed with the fluent API of our rich Iterator.

I know that it was hard work to provide all the implementations. This is really a drawback for maintainers of the collection hierarchy. If you don't have any objections regarding my suggestion, it would be great if you could change your PR accordingly.

Thanks
Daniel

**Comment by sleepytomcat** [comment]
Thank you @danieldietrich for detailed explanation! Let me follow-up with the code change.

**Comment by sleepytomcat** [comment]
I've changed the return type of `Traversable<>.unzip(Function, Function)` to `Tuple2<Iterator<T1>, Iterator<T2>`, and removed all the unnecessary container-specific code. Once the code looks good, I'll take another step to remove previous implementation of `unzip(Function<T, Tuple2<T1,T2>>)`.

**Comment by sleepytomcat** [comment]
@danieldietrich ,
```
// ...
final Iterator<T1> iter1 = map(unzipper1);
final Iterator<T2> iter2 = map(unzipper2);          
return Tuple.of(iter1, iter2);
```

—  I do not think `Iterator.unzip(Function, Function)` implementation would work this way. My reasoning: iterator is mutable, so traversing over, say, first Iterator `_1` in the resulting `Tuple2` will affect traversing of the iterator in `_2`.

Can you please share your perspective on this?

**Comment by sleepytomcat** [comment]
There's what I observe under debugger with said `Iterator.unzip(Function, Function)` implementation:

image

Please note empty `a_2` - my explanation this is because `a_1` iterator was already traversed, and there's no more data yielded by `a_2`.

Do you think this explanation may seem reasonable?

**Comment by danieldietrich** [comment]
> I do not think Iterator.unzip(Function, Function) implementation would work this way

You are absolutely right - it was a fast shot without really thinking about it. If we can't re-use an existing iterator (one of those internals in the Iterator interface), we have to write a new one that is able to feed two iterators. (Something like `tee` in bash).

Let be think about it, I will come back with a suggestion.

**Comment by danieldietrich** [comment]
Sadly, there is no chance to do it without allocating memory. This is the easiest way to do it (I did not test the code in an editor):

```java
    @Override
    default <T1, T2> Tuple2<Iterator<T1>, Iterator<T2>> unzip(
            Function<? super T, ? extends T1> unzipper1, Function<? super T, ? extends T2> unzipper2) {
        Objects.requireNonNull(unzipper1, "unzipper1 is null");
        Objects.requireNonNull(unzipper2, "unzipper2 is null");
        if (!hasNext()) {
            return Tuple.of(empty(), empty());
        } else {
            final Stream<T> stream = Stream.ofAll(() -> this);
            final Iterator<? extends T1> iter1 = stream.iterator().map(unzipper1);
            final Iterator<? extends T2> iter2 = stream.iterator().map(unzipper2);
            return Tuple.of(iter1, iter2);
        }
    }
```

Could you do this for Iterator.unzip und Iterator.unzip3, please?

**Comment by sleepytomcat** [comment]
Will do, thank you @danieldietrich

**Comment by sleepytomcat** [comment]
@danieldietrich , side question: from the `Iterator` class description, and from code implementation, it's not clear to me — generally speaking, is there a legitimate way to create a copy of an iterator instance? Or, they are inherently non-copyable? (I think copying an iterator instance could help us avoid memory allocation while traversing Tuple's _1, _2).

**Comment by sleepytomcat** [comment]
@danieldietrich , I've updated the code of `Tuple2<Iterator<T1>, Iterator<T2>> unzip(Function, Function)` according to your recommendations. `gradlew check` is successful now. Please kindly review and let know if it looks Ok. Once we're there, I will..
1) ...update `unzip3(Function<T, Tuple3>`) to `unzip3(Function<T, T1>, Function<T, T2>, Function<T, T2>)` — as you suggested;
2) ...add necessary tests coverage;
3) ...remove old `unzip` method using `Function<T, Tuple>` arguments, along with their tests.

Please let know if it looks Ok to you.

**Comment by danieldietrich** [comment]
@sleepytomcat

> side question (...) is there a legitimate way to create a copy of an iterator instance?

Currently there exists not way because Iterator is stateful (other than Vavr collections, which are persistent). However, the idea is tempting. I could imagine that an Iterator could have a copy() aka clone() aka duplicate() method that creates a new Iterator instance by copying the state. This might involve recursively copying wrapped Iterator instances. I will think a bit about it, but it is out of scope for this PR.

**Comment by danieldietrich** [comment]
@sleepytomcat Thank you, please go on with 1)-3) like you suggested!
Great that you updated the links to the methods in the javadoc at the top of the file! 👍🏼

File: src/main/java/io/vavr/collection/Traversable.java
Comment (sleepytomcat) [inline src/main/java/io/vavr/collection/Traversable.java:1635]: @danieldietrich , should this be 
`... Tuple3<? extends Iterator<T1>, ? extends Iterator<T2>, ? extends Iterator<T3>> unzip3(...) `
or
`... Tuple3<Iterator<T1>, Iterator<T2>, Iterator<T3>> unzip3(...)`?

Same question goes to `unzip(...)` above, line 1614.

**Comment by sleepytomcat** [comment]
@danieldietrich , the items (1)-(3) completed, please review the code and share your feedback. Thank you!

**Comment by sleepytomcat** [comment]
I think I messed up with pulling upstream changes. Fixing.

File: src/main/java/io/vavr/collection/Traversable.java
Comment (danieldietrich) [reply src/main/java/io/vavr/collection/Traversable.java:1635]: Please do it like this (for return types in general):

```java
Tuple3<Iterator<T1>, Iterator<T2>, Iterator<T3>>
```

Thank you!

**Review by danieldietrich — CHANGES_REQUESTED** [review]
Awesome, thank you!
I like it to remove all the code while adding more flexibility at the same time.
Please adjust the return type(s) like commented, then we are able to merge this beauty!

**Comment by sleepytomcat** [comment]
Return type changed to `Tuple3<Iterator<T1>, Iterator<T2>, Iterator<T3>>` as suggested. Thank you!

File: src/test/java/io/vavr/collection/AbstractTraversableTest.java
Comment (sleepytomcat) [inline src/test/java/io/vavr/collection/AbstractTraversableTest.java:2436]: @danieldietrich , I keep using `.toArray()` to 'materialize' data and compare `Iterators` in unit tests. Is there a better way to compare two iterators?

Comment (danieldietrich) [reply src/test/java/io/vavr/collection/AbstractTraversableTest.java:2436]: I think this is ok!

**Review by danieldietrich — APPROVED** [review]
Great, many thanks for this!

**Pull request merged**
